### PR TITLE
Emulator pinning

### LINKED ISSUE #27: allow pinning of emulator threads
libvirt should allow pinning of emulator threads to chosen host CPUs.

### PR DESCRIPTION
**Pull request by praveen-pk**
Fixes #27.

Only the last commit (ch_driver: Callbacks for emulator threadinfo and pinning) is relevant to this PR. All remaining commits are from https://github.com/cloud-hypervisor/libvirt/pull/26 PR.

### CODE REVIEW

**Comment by praveen-pk** [comment]
Following are the tests I ran:

```
$ virsh_ch dumpxml ch_net
<domain type='ch' id='199044'>
  <name>ch_net</name>
  <uuid>6fd72f2b-0f35-431a-9a04-4fcb351957fc</uuid>
  <title>Test Cloud Hypervisor Networking</title>
  <description>Network</description>
  <memory unit='KiB'>4194304</memory>
  <currentMemory unit='KiB'>4194304</currentMemory>
  <vcpu placement='static'>4</vcpu>
  <iothreads>3</iothreads>
  <cputune>
    <emulatorpin cpuset='4-5'/>
  </cputune>
  <resource>
    <partition>/machine</partition>
  </resource>
  <os>
    <type arch='x86_64'>hvm</type>
    <kernel>libvirt/vm/hypervisor-fw</kernel>
    <boot dev='hd'/>
  </os>
  <clock offset='utc'/>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <devices>
    <disk type='file' device='disk'>
      <driver name='qemu' type='raw' iothread='1'/>
      <source file='libvirt/vm/bionic-server-cloudimg-amd64.raw'/>
      <target dev='vda' bus='virtio'/>
    </disk>
    <disk type='file' device='disk'>
      <driver name='qemu' type='raw' iothread='2'/>
      <source file='libvirt/vm/cloud-config.raw'/>
      <target dev='vdb' bus='virtio'/>
    </disk>
    <interface type='ethernet'>
      <mac address='00:11:22:33:44:55'/>
      <ip address='192.168.4.2' family='ipv4' prefix='24'/>
      <target dev='ich0'/>
      <model type='virtio'/>
      <driver queues='4' rx_queue_size='256' tx_queue_size='256' iommu='on'/>
    </interface>
    <rng model='virtio'>
      <rate bytes='1234' period='2000'/>
      <backend model='random'>/dev/random</backend>
    </rng>
  </devices>
</domain>

$ virsh_ch emulatorpin ch_net
emulator: CPU Affinity
----------------------------------
       *: 4-5
	   
$ cat /sys/fs/cgroup/cpuset/machine.slice/machine-ch\\x2d199044\\x2dchnet.scope/emulator/tasks
199044
199045
199046
199062

$ ps -aeL | grep 199044
 199044  199044 pts/2    00:00:00 cloud-hyperviso
 199044  199045 pts/2    00:00:00 vmm
 199044  199046 pts/2    00:00:00 http-server
 199044  199058 pts/2    00:00:10 vcpu0
 199044  199059 pts/2    00:00:00 vcpu1
 199044  199060 pts/2    00:00:00 vcpu2
 199044  199061 pts/2    00:00:00 vcpu3
 199044  199062 pts/2    00:00:00 signal_handler
 199044  199063 pts/2    00:00:00 virtio_blk

$ cat /sys/fs/cgroup/cpuset/machine.slice/machine-ch\\x2d199044\\x2dchnet.scope/emulator/cpuset.cpus
4-5

$ virsh_ch emulatorpin --domain ch_net --cpulist 1,7
$ cat /sys/fs/cgroup/cpuset/machine.slice/machine-ch\\x2d199044\\x2dchnet.scope/emulator/cpuset.cpus
1,7

$ virsh_ch emulatorpin --domain ch_net
emulator: CPU Affinity
----------------------------------
       *: 1,7
```

**Comment by bryteise** [comment]
@praveen-pk I'm a bit behind sorry. I'll try to get a review out for this PR this week though. Do you mind updating the merge conflicts first though?

**Comment by praveen-pk** [comment]
> mind updating the merge

Just rebased :)

File: src/ch/ch_driver.c [hunk 1]
```diff
@@ -1325,6 +1325,158 @@ chDomainPinVcpu(virDomainPtr dom,
...
+    if (!(vm = chDomObjFromDomain(dom)))
+        goto cleanup;
+
+    if (virDomainGetEmulatorPinInfoEnsureACL(dom->conn, vm->def) < 0)
+        goto cleanup;
+
+    if (!(def = virDomainObjGetOneDefState(vm, flags, &live)))
+        goto cleanup;
+
+    if (live){
```
Comment (bryteise) [inline src/ch/ch_driver.c]: Nit, should have a space like `) {`.

File: src/ch/ch_driver.c [hunk 2]
```diff
@@ -1325,6 +1325,158 @@ chDomainPinVcpu(virDomainPtr dom,
...
+
+            if (chSetupCgroupCpusetCpus(cgroup_emulator, pcpumap) < 0) {
+                virReportError(VIR_ERR_OPERATION_INVALID, "%s",
+                               _("failed to set cpuset.cpus in cgroup"
+                                 " for emulator threads"));
+                goto endjob;
+            }
+        }
+
+        if (virProcessSetAffinity(vm->pid, pcpumap) < 0)
```
Comment (bryteise) [inline src/ch/ch_driver.c:1445]: This has changed in the upstream version and reminds me we'll need to take a look at rebasing the ch branch with all these changes soon.

**Comment by bryteise** [comment]
Just a slight nit and otherwise this is should be merged.

File: src/ch/ch_driver.c (hunk 1 above)
Comment (praveen-pk) [reply src/ch/ch_driver.c]: fixed!

File: src/ch/ch_driver.c (hunk 2 above)
Comment (praveen-pk) [reply src/ch/ch_driver.c:1445]: Makes sense @bryteise. Let us address this in the next rebase from upstream.

**Pull request merged**